elements/lisk-transaction-pool/src/transaction_list.ts

```diff
@@ -0,0 +1,249 @@
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+import { EventEmitter } from 'events';
+
+import { MinHeap } from './min_heap';
+import { Transaction } from './transaction_pool';
+
+export interface TransactionListOptions {
+	readonly maxSize?: number;
+	readonly minReplacementFeeDifference?: bigint;
+}
+
+const DEFAULT_MAX_SIZE = 64;
+const DEFAULT_REPLACEMENT_FEE_DIFF = BigInt(0);
+
+export const EVENT_TRANSACTION_ADDED = 'EVENT_TRANSACTION_ADDED';
+export const EVENT_TRANSACTION_REMOVED = 'EVENT_TRANSACTION_REMOVED';
+
+export class TransactionList {
+	public readonly address: string;
+	public readonly events: EventEmitter;
+
+	private _processable: Array<bigint>;
+	private readonly _transactions: { [nonce: string]: Transaction };
+	// Value is not needed here because it is stored separately in the _transactions
+	private readonly _nonceHeap: MinHeap<undefined, bigint>;
+	private readonly _maxSize: number;
+	private readonly _minReplacementFeeDifference: bigint;
+
+	public constructor(address: string, options?: TransactionListOptions) {
+		this.address = address;
+		this.events = new EventEmitter();
+		this._transactions = {};
+		this._nonceHeap = new MinHeap<undefined, bigint>();
+		this._processable = [];
+		this._maxSize = options?.maxSize ?? DEFAULT_MAX_SIZE;
+		this._minReplacementFeeDifference =
+			options?.minReplacementFeeDifference ?? DEFAULT_REPLACEMENT_FEE_DIFF;
+	}
+
+	public get(nonce: bigint): Transaction | undefined {
+		return this._transactions[nonce.toString()];
+	}
+
+	public add(incomingTx: Transaction, processable: boolean = false): boolean {
+		const existingTx = this._transactions[incomingTx.nonce.toString()];
+		// If the same nonce already exist in the pool try to replace
+		if (existingTx) {
+			// If the fee is lower than the original fee + replacement, reject
+			if (incomingTx.fee < existingTx.fee + this._minReplacementFeeDifference) {
+				return false;
+			}
+			// Mark this and all subsequenct nonce unprocessable
+			this._demoteAfter(incomingTx.nonce);
+			this._transactions[incomingTx.nonce.toString()] = incomingTx;
+			this.events.emit(EVENT_TRANSACTION_REMOVED, {
+				address: this.address,
+				id: incomingTx.id,
+			});
+
+			return true;
+		}
+		// If the size exceeds, remove the largest nonce
+		const highestNonce = this._highestNonce();
+		if (
+			this._nonceHeap.count >= this._maxSize &&
+			incomingTx.nonce > highestNonce
+		) {
+			return false;
+		}
+
+		this._transactions[incomingTx.nonce.toString()] = incomingTx;
+		this._nonceHeap.push(incomingTx.nonce, undefined);
+		// If this transaction is processable and it is the first one in the list
+		if (processable && this._processable.length === 0) {
+			this._processable.push(incomingTx.nonce);
+		}
+
+		// If the size exceeds, remove the largest nonce transaction
+		if (this._nonceHeap.count > this._maxSize) {
+			this.remove(highestNonce);
+		}
```

Comment:
maybe it better if we remove a transaction first when its reaches its `_maxSize ` so that we make sure at no given time the list size exceeds the `_maxSize `. So checking `if (this._nonceHeap.count >= this._maxSize) {` before `this._transactions[incomingTx.nonce.toString()] = incomingTx;`

Reply:
In case like account has nonce: 0, and in pool if it has [40, 41 ...125], and if nonce: 1 comes in, rather than rejecting, maybe it's better to remove 125?

Reply:
But if nonce `1` comes in, with below condition nonce `125` will be removed
```typescript
if (this._nonceHeap.count >= this._maxSize) {
			this.remove(highestNonce);
		}
```
and now, we can add 1,
```typescript
this._transactions['1'] = incomingTx;
this._nonceHeap.push(incomingTx.nonce, undefined);
// If this transaction is processable and it is the first one in the list
if (processable && this._processable.length === 0) {
   this._processable.push(incomingTx.nonce);
  }
```

Reply:
In this case, we will never have the list above `_maxSize`.

Reply:
ah you mean just move up the condition?

Reply:
yes :)